### PR TITLE
Fix array bounds error with EquivalentLayer window and shading surface

### LINKED ISSUE #10491: Array bounds error with EquivalentLayer window and shading surface
Issue overview
--------------
Stumbled onto this while developing a unit test for interior solar distribution with EquivalentLayer windows. Any file with EquivalentLayer windows and a shading surface will hit an array bounds error in `InitEquivalentLayerWindowCalculations`.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 24.2-https://github.com/NREL/EnergyPlus/commit/e0b62ad8d5ee9b0aa985b20bc21d38f591c1b32e

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file attached:
[EquivalentLayerWindow-AddShadingSurface.idf.txt](https://github.com/NREL/EnergyPlus/files/15201405/EquivalentLayerWindow-AddShadingSurface.idf.txt)

 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #10491

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mjwitte** [comment]
New unit test fails before fix is impelemented.
image

File: src/EnergyPlus/WindowEquivalentLayer.cc [hunk 1]
```diff
@@ -151,16 +146,15 @@ void InitEquivalentLayerWindowCalculations(EnergyPlusData &state)
...
         if (!state.dataConstruction->Construct(ConstrNum).TypeIsWindow) continue;
         if (!state.dataConstruction->Construct(ConstrNum).WindowTypeEQL) continue; // skip if not equivalent layer window
 
         SetEquivalentLayerWindowProperties(state, ConstrNum);
 
     } //  end do for TotConstructs
 
-    for (SurfNum = 1; SurfNum <= state.dataSurface->TotSurfaces; ++SurfNum) {
-        if (!state.dataConstruction->Construct(state.dataSurface->Surface(SurfNum).Construction).TypeIsWindow) continue;
+    for (int SurfNum : state.dataSurface->AllHTWindowSurfaceList) {
```
Comment (mjwitte) [inline src/EnergyPlus/WindowEquivalentLayer.cc:157]: In develop, this loops over all surfaces, including shading surfaces where `.Construction=0` causing an array bounds error. The fix reduces the loop to just window surfaces.

File: tst/EnergyPlus/unit/SolarShading.unit.cc [hunk 2]
```diff
@@ -5534,21 +5510,378 @@ TEST_F(EnergyPlusFixture, SolarShadingTest_CalcInteriorSolarDistribution)
...
+    EXPECT_NEAR(0.646, state->dataSurface->SurfWinGlTsolBmBm(windowSurfNum), 0.01);
+    EXPECT_NEAR(0.0, state->dataSurface->SurfWinGlTsolBmBm(windowSurfNum2), 0.01);
+    EXPECT_NEAR(0.0, state->dataHeatBalSurf->SurfWinInitialBeamSolInTrans(windowSurfNum), 0.01);
+    EXPECT_NEAR(0.0, state->dataHeatBalSurf->SurfWinInitialBeamSolInTrans(windowSurfNum2), 0.01);
+    EXPECT_NEAR(0.0, state->dataHeatBalSurf->SurfWinInitialDifSolInTrans(windowSurfNum), 0.01);
+    EXPECT_NEAR(1.2, state->dataHeatBalSurf->SurfWinInitialDifSolInTrans(windowSurfNum2), 0.01);
+}
+TEST_F(EnergyPlusFixture, SolarShadingTest_CalcInteriorSolarDistribution_EQL)
+{
+    // Test for equivalent layer window model
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/SolarShading.unit.cc:5573]: Clone existing unit test `SolarShadingTest_CalcInteriorSolarDistribution` and change window type to EquivalentLayer.

**Review by mjwitte — COMMENTED** [review]
Walkthru . . .

**Comment by mjwitte** [comment]
The defect file https://github.com/NREL/EnergyPlus/files/15201405/EquivalentLayerWindow-AddShadingSurface.idf.txt fails with develop, runs to completion here.

**Comment by mjwitte** [comment]
CI is all green, this is ready for review.

File: src/EnergyPlus/WindowEquivalentLayer.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/WindowEquivalentLayer.cc:157]: Nice that all those pre-set lists are in place so it's an easy change like this.

File: tst/EnergyPlus/unit/SolarShading.unit.cc (hunk 2 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/SolarShading.unit.cc:5573]: Maybe it would have been nicer to not have a near duplicate, but it's fine really.

**Review by Myoldmopar — APPROVED** [review]
Localizing a couple variables, fixing the array access, and adding a unit test to support the changes, no problem here.

**Comment by Myoldmopar** [comment]
CI is indeed all happy, this can merge now.  I'm not sure where @nrel-bot-3 is at....I am concerned I may have to go check on him.  ....  Anyway, thanks @mjwitte

**Pull request merged**